### ISSUE TITLE
Two instances of styled-components conflict when loaded on the same page

### ISSUE THREAD

**Issue by ryanheathers** [styled-components/styled-components#1032]
## Background
I develop a React app that gets installed as a 3rd-party plugin. That means multiple instances of my app may be running on a single page at any time. Along the way I've run into issues with how styled-component's shares the generated hashed css class amongst components on the page. If the same component is used twice, even if they are part of different React apps, they share the same class. This inevitably leads to style conflicts and specificity issues. [This issue](https://github.com/styled-components/styled-components/issues/633) has more background.

I've attempted - and largely succeeded - in working around this issue by the use of a style helper that wraps each styled-component in an ID selector. That way my generated styles look something like this:

```
#carousel-f75e1d05 .jaaqKx,
#carousel-f75e1d05-modal .jaaqKx {
    color: inherit;
    fill: currentColor;
    width: inherit;
    height: inherit;
}
```

Here is my helper function:

```
// Note: I've previously added a rootId UUID to my theme via code not shown here

export function rootIdWrapper(cssToUse) {
  return (props) => {
    return css`
      #${props.theme.rootId} &,
      #${props.theme.rootId}-modal & {
        ${cssToUse}
      }
    `;
  };
}
```

And here is how I use it:
```
const Svg = styled.svg`
  ${rootIdWrapper(css`
    color: inherit;
    fill: currentColor;
    width: inherit;
    height: inherit;
  `)}
`;
```

While it would be amazing to have first-class support for this kind of namespacing in styled-components itself, this slightly laborious workaround solves the issue.

## Expected Behavior
The styles for each of my app instances will be properly added to the page at run time.

## Actual Behavior
This code works great, that is, **when using styled-components v2.0.0-13.**

If I upgrade to a later version of styled-components, or revert to the base v2.0.0 version, I get buggy behavior:
- If **only one** of my React apps is loaded on the page, things are fine
- However if two or more instances are loaded, only the styles from one of the instances are loaded into the `<head>`. The other styles do not appear on the page at all.

## Steps to reproduce
It's difficult to provide an easily shareable replication of the issue since it involves concurrently running two React instances on a page. In my troubleshooting, I have produced a minimal example that uses the exact same code with the only difference being the version of styled-component's being used. **The version running 2.0.0-13 works great, any other 2.x version exhibits buggy behavior.**

Appreciate any insights out there on what the issue might be!

**Comment by philpl** [comment]
We currently don't support duplicated instances of styled-components. Is there any reason why you need this and can't share it across the bundle/app?

**Comment by ryanheathers** [comment]
> Is there any reason why you need this and can't share it across the bundle/app?

It's very important to our use case to have each app's dependencies fully contained inside itself. This prevents a host of dependency conflict issues.

I'm a tad confused why you say styled-component's doesn't support multiple instances of itself on a page. Clearly it did, at one point, because both the 1.x version that I used to use and v2.0.0-13 work great with having multiple instances of styled-component's loaded on the page. Something must have changed in your architecture? Also I would think you would want to support this use case, because otherwise styled-components cannot be safely used in any code that will be distributed into environments that the app author doesn't fully control.

**Comment by ryanheathers** [comment]
@philpl after reading [the issue](https://github.com/styled-components/styled-components/issues/1011) you linked, it sounds like that is referring to having two instances of styled-components in the same bundle. That is not the case for my apps. Each app has one and only instance of styled-components per bundle.

My issue (presumably) has more to do with whether styled-components affects the globals on the webpage and whether it conflicts with other instances of styled-components that may be trying to update the `<head>` at the same time. Multiple instances of styled-components on a page has seemed to work - outside of the class specificities issues I've detailed before and am trying to workaround now - unless something has changed in your architecture in recent releases.

**Comment by philpl** [comment]
@ryanheathers It comes down to the same issue: Running multiple instances of styled-components in a single app/page (whether it's bundled doesn't matter).

SC has no way of knowing whether there are multiple instances right now, and we make the assumption that it's undesirable anyway, since then you'd deploy multiple times the size of what you actually need to send to the user's browser.

What prevents you from sharing SC across bundles? Like with using `externals` on webpack?

Or `peerDependencies` when it's all in one build (multiple bundles per chunk splitting and vendoring with e.g. CommonsChunkPlugin, or even code splitting with dynamic imports)

**Comment by ryanheathers** [comment]
@philpl I appreciate the dialogue.

> What prevents you from sharing SC across bundles? Like with using externals on webpack?

We deliberately don't want to rely on our environment to supply our dependencies. That opens us up to a host of potential bugs if the environment doesn't supply us what we expect or doesn't supply it in time. Race conditions, version mismatches, etc etc. It's actually how we used to load in certain dependencies and it was a nightmare.

And even if we figure out how to load the one universal copy of styled-components that our 3rd party apps need, that doesn't solve the issue for any other 1st party or 3rd party apps running on the page that may want their own version of styled-components. Our version will conflict with theirs.

I'll reiterate my earlier words:
_Also I would think you would want to support this use case [of multiple instances], because otherwise styled-components cannot be safely used in any code that will be distributed into environments that the app author doesn't fully control._ 

This is why jQuery.noConflict() exists. React now also plays nicely with multiple instances on the same page because they recognized the importance of this use case. More discussion on that [here](https://github.com/facebook/react/issues/1939) and [here](https://github.com/facebook/react/issues/2402).

> SC has no way of knowing whether there are multiple instances right now, and we make the assumption that it's undesirable anyway, since then you'd deploy multiple times the size of what you actually need to send to the user's browser.

In addition to my thoughts above, (a) it seems like SC used to know or at least didn't conflict and (b) we're willing to ship the same dependency because the benefits of self-contained dependencies outweighs the cost of shipping more code over the wire.

**Comment by philpl** [comment]
> for any other 1st party or 3rd party apps running on the page that may want their own version of styled-components

Optimally yes, that should be possible, but I'm assuming you're talking about *one app* :wink: and I don't really see how it's best practice to have duplicated modules (especially some on the bigger side), in a single app.

It is possible for dependencies to use the same lib, as long as they're part of the build process. It's really all about the build process that should make this possible.

It's not like we'd ever discuss: I have three sub-apps on this page, so I want to have React loaded three times. That's insanity! Easy build processes but insanely expensive.

I think we'd want to support this eventually, but I'm wary of stuff like this, as it's usually not important or leads to other issues due to workarounds and special setups like this. I agree, that more uncommon use-cases need to be covered, but it's tricky if it opens up the door for practices that we don't recommend, or don't test for.

>  it seems like SC used to know or at least didn't conflict

We didn't use to have rehydration. And adding this makes SC look for an existing tag. We can probably fix this, but we'll need to solve other issues together with it as well. So we'll need to look for example for how to render multiple SSR style tags. cc @geelen

> we're willing to ship the same dependency because the benefits of self-contained dependencies outweighs the cost of shipping more code over the wire.

But with great tools like Webpack, is it really such a pain anymore? :)

**To sum it up:** Sorry, for asking a lot of times whether you need this, but you're the first one who expresses such a different use-case, and I can't promise you that it will move up the priority list unless more users need it. Especially since right now a lot of styled-components v3 work takes priority for me.

cc @mxstbr; any ideas for quick fixes? Or any idea on whether we might want to move this up our priority list, and can scope some smaller fix/solution quickly?

**Comment by philpl** [comment]
@ryanheathers If you want to try sth out in your environment really quickly, can you try this out?

https://github.com/styled-components/styled-components/blob/master/src/models/BrowserStyleSheet.js#L125

`const nodes = document.querySelectorAll(`[${SC_ATTR}]`)`

Here we're searching for all style tags that might be rendered on the server side, but we're not filtering for non-local ones. It might just work to extend the query to include another condition:

`[${LOCAL_ATTR}="false"]`

**Comment by ryanheathers** [comment]
> Optimally yes, that should be possible, but I'm assuming you're talking about one app 😉 and I don't really see how it's best practice to have duplicated modules (especially some on the bigger side), in a single app.

No, I'm talking about separate React apps, each of which uses it's own copy of styled-components.

If it makes it easier, ignore that fact that my app might be loaded multiple times on a webpage. Instead consider a scenario where 3 completely different React apps are used on the same webpage: App A, App B, App C. Each of the apps, unbeknownst to each other, loads a version of styled-components. What happens currently? Do the three copies of styled-components that are present on the page conflict in some fashion?

It sounds like maybe your rehydration code is now causing a conflict, where in older versions of styled-components that was no longer the case.

**Comment by geelen** [comment]
Hey sorry I'm back from holidays and just catching up on the issue tracker and I think I know what's going on.

Basically, Styled Components now includes rehydration, which means it will scan anything in the DOM to see if there are `<style>` tags generated by SC it should reuse. Of course, this means that if two copies of SC boot up, and the first renders styles before the second boots, and... I think... you're not running the babel transform to get deterministic IDs, the second app will generate the same classnames for some styles, think it needs to rehydrate, and you'll get these kinds of conflicts.

With the babel transform in place, each component gets an ID partially based on the pathname of the file its written in, which then gets fed into the hash. So you _should_ see fewer conflicts if you use that. Can you try and confirm whether it solves it?

There's also the case that you're sharing component files between these apps, so that won't work. The solution there is to allow SC to boot in a "don't rehydrate" mode. This is the sort of config (along with turning off autoprefixing) that I've been trying to find a nice place to introduce. If rehydration was optional, then each app would inject and maintain its own style tags, just like v1.

I'll be able to get that worked on and out at CampJS next weekend, I reckon, if not before.

**Comment by ryanheathers** [comment]
@geelen thanks for the thoughts.

> With the babel transform in place

I assume you're talking about [this plugin](https://github.com/styled-components/babel-plugin-styled-components. )? Unfortunately running that transform with styled-components v2.1.1 still results in the same conflicts. Note that I'm configuring the plugin with just the default `.babelrc` config of ```plugins: ["styled-components"]``` so if there is some other config I should try please let me know.

> There's also the case that you're sharing component files between these apps, so that won't work. 

This is indeed the case.

> The solution there is to allow SC to boot in a "don't rehydrate" mode. This is the sort of config (along with turning off autoprefixing) that I've been trying to find a nice place to introduce. If rehydration was optional, then each app would inject and maintain its own style tags, just like v1.

This sort of config sounds amazing. Happy to test it out as soon as it's available. Appreciate you tackling this!

**Comment by wmdmark** [comment]
Our company is also running into this problem. We have a core "UI" built with styled-components used by both our internal team and by external development partners. 

We also use styled-components in our other app packages and then things get borked when we import from our UI library. 

Does anyone know of a temporary solution while this issue is being worked on (ideally that works without ejecting create-react-app since we want to keep things simple for our partner developers)?

**Comment by philpl** [comment]
@wmdmark the temporary and in fact proper solution is to share the styled-components library. If you are certain that you have completely separate bundles on one single page, and are using webpack, give externals a try.

**Comment by wmdmark** [comment]
@philpl what do you mean by share the styled-components library? Bundle it with the project? The UI lib is an npm package that can be used by other projects (internal and external). The issue is that when those projects use styled-components themselves they get borked as soon as the UI library is imported as a dependency. 

Seems like this is a pretty major issue for anyone trying to create npm packages that use styled-components or maybe I'm just misunderstanding something? 

FWIW, we had to downgrade `styled-components` back to 1.4 to get around this problem.

**Comment by philpl** [comment]
@wmdmark You have to make sure that the module is not duplicated. That's a basic problem with a lot of libraries, not exclusive to ours. Otherwise you also run the risk of duplicating the SC module that you're shipping to the client.

There are a couple of options to fix this:

- Make sure that in your shared npm packages you depend on SC using `peerDependencies`, not a normal one, to make deduplication easy
- If some packages (which are out of your control) have SC as a normal dependency, make sure that their version ranges overlap so npm/yarn can deduplicate them
- Deduplicate them on a webpack level
- Externalise them in webpack and load the minified UMD bundle separately

It's unlikely that we'll fix having multiple instances of SC on one page anytime soon. In fact, we'd like to warn about it instead. But if you'd like to start fixing it or help add the warning, check out this issue: https://github.com/styled-components/styled-components/issues/1011

**Comment by wmdmark** [comment]
@philpl  Thanks for the tips, but I don't think this addresses the issue we're having since styled-components is _not_ being duplicated in this scenario (our UI lib builds without styled-components bundled). However, when I import anything from the UI lib, it immediately breaks any local styled-components being used in the project. 

Is there anything else that would cause this besides duplication? Also seems odd to me that the issue goes away after downgrading to 1.4.

**Comment by philpl** [comment]
@wmdmark then your problem is not related to this issue, since it doesn't match its description of two React apps (i.e. bundles). Please create a separate one

**Comment by wmdmark** [comment]
@philpl OK, moved to #1076

**Comment by ryanheathers** [comment]
@geelen checking to see if there is movement on this issue. For now I am working around it by simply using an older version of styled-components, but obviously would love to use the latest and greatest. Really appreciate the work here!

**Comment by ypling** [comment]
@ryanheathers, I think your work around would help with my use case:

I want to use styled component in an existing project which has a lot global styles.
In order to stop being effected by any old styles. I added class name 'reset' to the container of my new UI components.
```
<div class="reset">
styled components goes in here.
</div>

.reset * {
all: unset;
}
```
So this reset will override SC generated styles without your work around.
But with sass I can easily wrap all styles in .reset{} to get higher specificity.
Also won't be a problem for inline style.

**Comment by prevostc** [comment]
Same issue here, I have a react widget that anyone can install with a google-analytics-like js tag. 

However, this widget conflicts with my customer apps that use styled components too (and it was quite hard to debug :D)
It turns out that there is a conflicts in the style tags. The first inserted style tag (my client's tag) gets kicked out by my widget style tag.

I would like to be able to set the SC_ATTR variable. Here are the possible fixes (IMHO):

- At build time, with an env var, like process.env.SC_ATTR
- At run time, with a new styled.config() function
- At run time, with a ConfigProvider

I can contribute if any of these solutions does fit.

**Comment by philpl** [comment]
This will either be solved by: https://github.com/styled-components/styled-components/pull/1102 or by https://github.com/styled-components/styled-components/issues/1011

Since the last comments have not been defining the problem further we can move on to the issue/PR above addressing some solutions instead. I'll go ahead and lock this issue, as use-cases or +1s are not going to be helpful. I hope the issue and PR above will sufficiently solve this, but I do believe so.

@geelen: Feel free to unlock of course if anything changes :)
